File: src/main/java/com/example/poc/ContentInterceptor.java
package com.example.poc;

import org.springframework.http.HttpRequest;
import org.springframework.http.client.ClientHttpRequestExecution;
import org.springframework.http.client.ClientHttpRequestInterceptor;
import org.springframework.http.client.ClientHttpResponse;

import java.io.IOException;

public class ContentInterceptor implements ClientHttpRequestInterceptor {

    @Override
    public ClientHttpResponse intercept(HttpRequest request, byte[] body, ClientHttpRequestExecution execution) throws IOException {

        var response = execution.execute(request, body);
        // Consume response body twice. Should work because the test enables buffered content
        System.out.println(response.getBody().readAllBytes());
        System.out.println(response.getBody().readAllBytes());

        return response;
    }
}

File: src/main/kotlin/com/example/poc/ContentInterceptor.kt
package com.example.poc

import org.springframework.http.HttpRequest
import org.springframework.http.client.ClientHttpRequestExecution
import org.springframework.http.client.ClientHttpRequestInterceptor
import org.springframework.http.client.ClientHttpResponse
import org.springframework.stereotype.Component

@Component
class ContentInterceptor: ClientHttpRequestInterceptor {
    override fun intercept(
        request: HttpRequest,
        body: ByteArray,
        execution: ClientHttpRequestExecution,
    ): ClientHttpResponse {
        val response = execution.execute(request, body)
        // Read response twice. This should work when buffering is enabled
        println(response.body.readAllBytes().decodeToString())
        println(response.body.readAllBytes().decodeToString())

        return response
    }
}
